cmd/dagger/exec.go

```diff
@@ -0,0 +1,56 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+
+	"github.com/spf13/cobra"
+)
+
+var execCmd = &cobra.Command{
+	Use:                   "exec [command]",
+	Aliases:               []string{"x"},
+	DisableFlagsInUseLine: true,
+	Long:                  "Executes the specified command in a Dagger session\n\nDAGGER_SESSION_PORT will be injected automatically.",
+	Short:                 "Executes a command in a Dagger session",
+	Example: `
+dagger exec -- sh -c 'curl \
+-H "content-type:application/json" \
+-d "{\"query\":\"{container{id}}\"}" \
+http://localhost:$DAGGER_SESSION_PORT/query'`,
+	Run:  Exec,
+	Args: cobra.MinimumNArgs(1),
+}
+
+func Exec(cmd *cobra.Command, args []string) {
+	ctx := context.Background()
+	if err := setupServer(ctx); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+
+	listening := make(chan string)
+	go func() {
+		// allocate the next available port
+		l, err := net.Listen("tcp", "localhost:0")
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+			os.Exit(1)
+		}
+		listening <- fmt.Sprintf("%d", l.Addr().(*net.TCPAddr).Port)
+		http.Serve(l, nil)
+	}()
+
+	listenPort := <-listening
+	os.Setenv("DAGGER_SESSION_PORT", listenPort)
```

Comment:
This shouldn't affect SDK's since the idea of setting `DAGGER_SESSION_PORT` is so the wrapped process can use it to connect to `localhost:$PORT` to reach the API. For example: 
```
dagger exec -- sh -c 'curl -v \
-H "content-type:application/json" \
-d "{\"query\":\"{container{id}}\"}" \
http://localhost:$DAGGER_SESSION_PORT/query'
```

Reply:
Oh my understanding was that `dagger exec` was meant to wrap any command that uses a dagger sdk (so that e.g. everything wrapped by the command would be in the same session) as opposed to being meant to wrap literally anything, but I may be out of sync with the latest ideas around this.